src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java

```diff
@@ -0,0 +1,184 @@
+package com.internet.shop.dao.jdbc;
+
+import com.internet.shop.dao.UserDao;
+import com.internet.shop.exeptions.DataProcessingException;
+import com.internet.shop.lib.Dao;
+import com.internet.shop.model.Role;
+import com.internet.shop.model.User;
+import com.internet.shop.util.DbConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Dao
+public class UserDaoJdbcImpl implements UserDao {
+    @Override
+    public Optional<User> findByLogin(String login) {
+        String query = "SELECT *, group_concat(roles.role_name SEPARATOR ',') AS role_titles"
+                        + " FROM users"
+                        + " JOIN users_roles ON users.user_id = users_roles.user_id"
+                        + " JOIN roles ON users_roles.role_id = roles.role_id"
+                        + " where users.login = ?";
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setString(1, login);
+            ResultSet sqlResult = preparedStatement.executeQuery();
+            if (sqlResult.next()) {
+                User user = parseRow(sqlResult);
+                return Optional.of(user);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get user by login = " + login, e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public User create(User user) {
+        String query = "INSERT INTO users (user_name, password, login)"
+                + " VALUES (?, ?, ?)";
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement =
+                    connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setString(1, user.getName());
+            preparedStatement.setString(2, user.getPassword());
+            preparedStatement.setString(3, user.getLogin());
+            preparedStatement.executeUpdate();
+            ResultSet resultKey = preparedStatement.getGeneratedKeys();
+            if (resultKey.next()) {
+                user.setId(resultKey.getLong(1));
+            }
+            addUserRoles(user, connection);
+        } catch (SQLException e) {
+            throw new DataProcessingException("User with login " + user.getLogin()
+                    + " has not been added", e);
+        }
+        return user;
+    }
+
+    private void addUserRoles(User user, Connection connection) throws SQLException {
+        for (Role role : user.getRoles()) {
```

Comment:
I have a comment from mentors - look on p8 from checklist for this HW. Cycle for should be after setting up connection otherwise you might create too many connections

Reply:
I got open connection - new connection hasn't created in this method